### PR TITLE
Fix docker auth

### LINKED ISSUE #21: OSX failes publishing to registry (and probably saving) when authentication required
Probably not a high priority for publish but I suspect save will be. OSX uses a keychain provider thingy...

### PR DESCRIPTION
**Pull request by lewismarshall**
- Adds auth to pull
- Fixes panic #24 

Also fixes #21

### CODE REVIEW

File: pkg/docker/save.go [hunk 1]
```diff
@@ -30,7 +30,15 @@ func Save(image string, dir string) error {
...
 		return (err)
 	}
-	events, err := cli.ImagePull(ctx, image, types.ImagePullOptions{})
+	var ipo types.ImagePullOptions
+	// Load auth details from .docker config
+	authStr := GetAuth(image)
+	if len(authStr) > 0 {
+		ipo = types.ImagePullOptions{RegistryAuth: authStr}
+	} else {
+		ipo = types.ImagePullOptions{}
```
Comment (gambol99) [inline pkg/docker/save.go]: i think this is doing the same thing as `var ipo types.ImagePullOptions` ... perhaps we can shrink the lines to 
```Go
var ipo types.ImagePullOptions
if len(authStr) > 0 { 
   ipo.RegistryAuth = GetAuth(image)
}
```

File: pkg/docker/config.go [hunk 2]
```diff
@@ -12,8 +12,11 @@ import (
 // GetConfig loads config for a given registry from the Docker config file
 func GetAuth(image string) string {
 	registry := strings.Split(image, "/")[0]
-	auths, _ := c.NewAuthConfigurationsFromDockerCfg()
-
+	auths, err := c.NewAuthConfigurationsFromDockerCfg()
+	if err != nil {
+		log.Printf("warning (errro getting authentication details: %s)", err)
```
Comment (gambol99) [inline pkg/docker/config.go]: s/errro/error/

File: pkg/docker/config.go [hunk 3]
```diff
@@ -12,8 +12,11 @@ import (
 // GetConfig loads config for a given registry from the Docker config file
```
Comment (gambol99) [inline pkg/docker/config.go]: method comment needs updating

File: pkg/docker/save.go (hunk 1 above)
Comment (lewismarshall) [reply pkg/docker/save.go]: Yep - works - declare = nil things, lib is happy.

File: pkg/docker/config.go (hunk 2 above)
Comment (lewismarshall) [reply pkg/docker/config.go]: Chrz, catching my dyslexic abilities.

File: pkg/docker/config.go (hunk 3 above)
Comment (lewismarshall) [reply pkg/docker/config.go]: yep nice.

**Comment by vinn946** [comment]
When docker config doesn't exist (on CentOS):
```
[vincentlam@td-dsp01 crown-high-elk]$ ../artefactor_linux_amd64 publish --docker-registry registry.dsp.internal --logs
...
Loading image from downloads/minio~minio~~RELEASE.2018-04-12T23-41-09Z.docker.tar.
2018/07/03 14:05:31 {"stream":"Loaded image: minio/minio:RELEASE.2018-04-12T23-41-09Z\n"}
ReTagging image as registry.dsp.internal/minio:RELEASE.2018-04-12T23-41-09Z.
pushing image registry.dsp.internal/minio:RELEASE.2018-04-12T23-41-09Z
2018/07/03 14:05:31 warning (errro getting authentication details: open /home/vincentlam/.dockercfg: no such file or directory)
Error: problem pushing image registry.dsp.internal/minio:RELEASE.2018-04-12T23-41-09Z to registry:Error response from daemon: Bad parameters and missing X-Registry-Auth: EOF
Usage:
  artefactor publish [flags]

Flags:
      --archive-dir string       a directory where artefacts exist to publish from (${ARTEFACTOR_ARCHIVE_DIR}) (default "downloads")
      --docker-registry string   where to publish images e.g. private-registry.local (${ARTEFACTOR_DOCKER_REGISTRY})
  -h, --help                     help for publish

Global Flags:
      --docker-images string   A whitespace seperated list of docker images (${ARTEFACTOR_DOCKER_IMAGES})
      --git-repos string       A whitespace seperated list of local or remote git repos (${ARTEFACTOR_GIT_REPOS})
      --logs                   Enable debug logs
      --web-files string       A whitespace seperated list of CSV's: url,filename,sha256,true (${ARTEFACTOR_WEB_FILES})

2018/07/03 14:05:31 problem pushing image registry.dsp.internal/minio:RELEASE.2018-04-12T23-41-09Z to registry:Error response from daemon: Bad parameters and missing X-Registry-Auth: EOF
```
Also with MacOS - expected to fail as it uses its keychain provider
```
vincent@Vincents-MacBook-Pro:~/temp/artefactor/crown-high-elk $ artefactor publish --docker-registry artifactory-internal.digital.homeoffice.gov.uk --logs
...
Loading image from downloads/quay.io~ukhomeofficedigital~elasticsearch-curator~~v0.3.0.docker.tar.
2018/07/03 15:05:45 {"stream":"Loaded image: quay.io/ukhomeofficedigital/elasticsearch-curator:v0.3.0\n"}
ReTagging image as artifactory-internal.digital.homeoffice.gov.uk/elasticsearch-curator:v0.3.0.
pushing image artifactory-internal.digital.homeoffice.gov.uk/elasticsearch-curator:v0.3.0
2018/07/03 15:05:45 warning (errro getting authentication details: Failed to read authentication from dockercfg)
Error: problem pushing image artifactory-internal.digital.homeoffice.gov.uk/elasticsearch-curator:v0.3.0 to registry:Error response from daemon: Bad parameters and missing X-Registry-Auth: EOF
Usage:
  artefactor publish [flags]

Flags:
      --archive-dir string       a directory where artefacts exist to publish from (${ARTEFACTOR_ARCHIVE_DIR}) (default "downloads")
      --docker-registry string   where to publish images e.g. private-registry.local (${ARTEFACTOR_DOCKER_REGISTRY})
  -h, --help                     help for publish

Global Flags:
      --docker-images string   A whitespace seperated list of docker images (${ARTEFACTOR_DOCKER_IMAGES})
      --git-repos string       A whitespace seperated list of local or remote git repos (${ARTEFACTOR_GIT_REPOS})
      --logs                   Enable debug logs
      --web-files string       A whitespace seperated list of CSV's: url,filename,sha256,true (${ARTEFACTOR_WEB_FILES})

2018/07/03 15:05:45 problem pushing image artifactory-internal.digital.homeoffice.gov.uk/elasticsearch-curator:v0.3.0 to registry:Error response from daemon: Bad parameters and missing X-Registry-Auth: EOF
```

**Comment by lewismarshall** [comment]
@vinn946 Re-opened and renamed issue #21 This issue just deals with the panic (unfortunately).

**Comment by lewismarshall** [comment]
@vinn946 I think it's worth getting this in, as tracking osx  issue #21 separately. @gambol99 dealt with your comments... review pls...

**Pull request merged**